book/layout.md

```diff
@@ -5,529 +5,632 @@ prev: html
+contains any number of blocks, or a single inline, while inline layout
+has no children. So for example, a document with a heading and two
+paragraphs would correspond to a tree like this:
+
+[^in-english]: In European languages. But not universally!
+
+![A tree of layout objects. The root is a block for the `<html>`
+    element; it has one child, a block for the `<body>` element; it
+    has three children, each blocks as well (for the `<h1>` and two
+    `<p>` elements); and each of those have a single child, an
+    inline.](/im/layout-modes.png)
+
+Finally: layout. With the layout tree created, how do we compute the
+size and position of each layout object? The general rule is that a
+block, like a paragraph, should take up as much horizontal room as it
+can, and should be tall enough to contain everything inside it. That
+means that a layout objects's width is based on its *parent*'s width,
+while its *height* is based on its *children*'s height:
+
+![The flow of information through layout. Width information flows down
+    the tree, from parent to child, while height information flows up,
+    from child to parent.](/im/layout-order.png)
+
+This suggests a step-by-step approach to layout. First, an element
+must compute its width, based on its parent's width. That makes it
+possible to lay out the children, which comes next. Finally, the
+children's heights are now available, so the element's height can be
+calculated.
+
+Let's now turn these concepts into code.
+
+::: {.further}
+Formally, computations on a tree like this can be
+described by an [attribute grammar](wiki-atgram). Attribute grammar
+engines analyze dependencies between different attributes to determine
+the right order to calculate each attribute.
+:::
+
+[wiki-atgram]: https://en.wikipedia.org/wiki/Attribute_grammar
+
+Layout modes
+============
 
-Well, that works for simple layouts, but it's pretty hard to draw a
-border without knowing how wide and tall to draw the border.
+Let's start with the scaffolding for layout modes. We'll have two,
+block and inline layout, and since what we've written now is basically
+inline layout let's rename `Layout` to `InlineLayout` and add a new
+`BlockLayout` class:
 
-So web browsers actually structure layout differently. In a browser,
-the job of layout is to produce a tree, called the *box tree*, so that
-each element has an associated box and each box is annotated with a
-size and a position. When you lay out text, you do so relative to the
-box the element is contained in.
+``` {.python}
+class InlineLayout:
+    # ...
 
-Our `layout` function doesn\'t do this yet, so let's fix it.
+class BlockLayout:
+    # ...
+```
 
-Let's start by defining a data structure to store an area where text
-can be laid out. I\'m going to call this a `Block`:
+These layout objects will be the nodes of the layout tree, and in fact
+some browsers call it a layout tree. To make it a tree, we'll want
```

Comment:
Remove this second part of the sentence? Not sure it adds value and avoids confusion.